### PR TITLE
Euclidean batch support

### PR DESCRIPTION
**Pull request by nicholas-leonard**
Includes unit tests. Optimized for the new non-contiguous copy kernels (https://github.com/torch/cutorch/pull/93).

### CODE REVIEW

File: Euclidean.lua [hunk 1]
```diff
@@ -32,36 +33,146 @@ function Euclidean:reset(stdv)
...
+   self._expand = self._expand or self.output.new()
+   self._expand2 = self._expand2 or self.output.new()
+   self._repeat = self._repeat or self.output.new()
+   self._repeat2 = self._repeat2 or self.output.new()
+   
+   local inputSize, outputSize = self.weight:size(1), self.weight:size(2)
+   
+   -- y_j = || w_j - x || = || x - w_j ||
+   if input:dim() == 1 then
+      self._input:reshape(input, inputSize, 1)
```
Comment (soumith) [inline Euclidean.lua]: reshape means a copy for sure. can we make a :viewAs out of this line by any chance?

Comment (soumith) [inline Euclidean.lua]: same with the input reshapes at the other lines

File: Euclidean.lua [hunk 2]
```diff
@@ -32,36 +33,146 @@ function Euclidean:reset(stdv)
...
+      if torch.type(input) == 'torch.CudaTensor' then
+         -- requires lots of memory, but minimizes cudaMallocs and loops
+         self._repeat2:resizeAs(self._expand2):copy(self._expand2)
+         self._repeat:add(-1, self._repeat2)
+      else
+         self._repeat:add(-1, self._expand2)
+      end
+      
+      self.output:norm(self._repeat, 2, 2)
+      self.output:resize(batchSize, outputSize)
```
Comment (soumith) [inline Euclidean.lua]: resize, or :view again?

**Comment by nicholas-leonard** [comment]
Oh my god this shatters my entire existence. All this time I assumed reshape did not perform a copy.

**Comment by nicholas-leonard** [comment]
Fixed.

**Comment by soumith** [comment]
thanks!

**Pull request merged**